[DOCS-422] Documenting HubAction API Reference

rajkaramchedu opened the pull request:
This is a document for HubAction API reference. The "hubaction-ref.rst" is the main document. 
@lbredeso @jimmyjames please review
/cc @bflorian @jodyalbritton @workingmonk 

Thanks for all your help!

Raj

lbredeso commented on ref-docs/hubaction-ref.rst:
```diff
@@ -0,0 +1,102 @@
+	Option   Description
+	======== ===========
+	callback Name of the callback method
+	type	 Type of LAN request. Allowed values are: LAN_TYPE_CLIENT, LAN_TYPE_SERVER. Default value is LAN_TYPE_CLIENT.
```
I did some double checking in the code, and this is a bit confusing, so I wanted to try to clarify what is going on behind the scenes:

The `protocol` here in `options` is as described below for `params` (LAN_PROTOCOL_TCP, LAN_PROTOCOL_UDP, with default of LAN_PROTOCOL_TCP).

lbredeso replied:
`protocol` still needs to be documented here for `options`, with possible values of LAN_PROTOCOL_TCP and LAN_PROTOCOL_UDP, and with a default of LAN_PROTOCOL_TCP

lbredeso commented on ref-docs/hubaction-ref.rst:
```diff
@@ -0,0 +1,102 @@
+	path 		Allowed values are any string of the form ``"/somepath"``. Default value is ``"/"``.
+	method 		Allowed values are ``"POST"``, ``"GET"``, ``"PUT"`` and ``"PATCH"``. Default value is ``"POST"``.
+	protocol 	Allowed values are LAN_PROTOCOL_TCP, LAN_PROTOCOL_UDP. Default value is LAN_PROTOCOL_TCP.
+	proto 		Default value is ``"HTTP/1.1"`` It is overridden with the *protocol* parameter above.
```
Developers are not able to pass a `proto` parameter to the `params` argument, so it should be omitted.

lbredeso commented on ref-docs/hubaction-ref.rst:
```diff
@@ -0,0 +1,102 @@
+	method 		Allowed values are ``"POST"``, ``"GET"``, ``"PUT"`` and ``"PATCH"``. Default value is ``"POST"``.
+	protocol 	Allowed values are LAN_PROTOCOL_TCP, LAN_PROTOCOL_UDP. Default value is LAN_PROTOCOL_TCP.
+	proto 		Default value is ``"HTTP/1.1"`` It is overridden with the *protocol* parameter above.
+	contentType	Allowed values are ``'application/json'``, ``'text/xml; charset="utf-8"'``.
```
Developers are not able to pass a 'contentType' parameter to the params argument, so it should be omitted. The Content-Type value is pulled directly from `headers`, which can be any valid HTTP/1.1 headers. If the Content-Type is not included in `headers`, we attempt to automatically set it using the following rules:

1. If `params.body` is a `Map`, set the Content-Type to be `application/json`
2. Otherwise, set the Content-Type to be `text/xml; charset="utf-8"`

lbredeso commented on ref-docs/hubaction-ref.rst:
```diff
@@ -0,0 +1,102 @@
+    // the below calledBackHandler() is triggered when the device responds to the sendHubCommand() with "device_description.xml" resource
+
+    void calledBackHandler(physicalgraph.device.HubResponse hubResponse) {
+    	log.debug "Entered calledBackHandler1()..."
```
I think this log statement should be "Entered calledBackHandler()..."

rajkaramchedu commented:
@lbredeso thanks for your detailed review. I fixed all you highlighted, namely: Fixed descriptions for "protocol", "Content-Type" in "headers", and deleted "proto" and "contentType" options. More important, we know that ``Map`` params.protocol currently only supports ``Protocol.LAN`` value, and since ``Protocol protocol`` gets its values from ``Map`` params.protocol, we are describing that for "Protocol.LAN" is the only allowed and default value for ``Protocol protocol`` parameter. I think this captures it, doesn't it?

rajkaramchedu commented:
@lbredeso Done the changes. Please review, thanks!
